File: Common/Engine/src/common/engine/components/game/GameEntryPoints.java
package common.engine.components.game;

import javafx.application.Platform;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Optional;

public final class GameEntryPoints extends ArrayList<GameEntryPoint> {

    public GameEntryPoints() {
    }

    public GameEntryPoints( GameEntryPoint... gameEntryPoints ) {
        Collections.addAll( this, gameEntryPoints );
    }

    public GameEntryPoint get(final String name ) {
        for ( GameEntryPoint g : this ) {
            if (g.getName().equals(name)) {
                return g;
            }
        }
        throw new IllegalArgumentException( "GameEntryPoint not found" );
    }

    @Override
    public boolean contains( final Object o ) {
        if ( o instanceof GameEntryPoint ) {
            final GameEntryPoint g        = (GameEntryPoint) o;
            final String         gameName = g.getName();
            for ( GameEntryPoint gameEntryPoint : this ) {
                if ( gameEntryPoint.getName().equals( gameName ) ) {
                    return true;
                }
            }
            return false;
        } else if (o instanceof String) {
            return this.contains((String) o);
        }
        return false;
    }

    public boolean contains( final String gameName ) {
        for ( GameEntryPoint g : this ) {
            if (g.getName().equals(gameName)) {
                return true;
            }
        }
        return false;
    }

    public void render( final int fps ) {
        for (GameEntryPoint game : this) {
            Platform.runLater( () -> {
                game.render( fps );
            });
        }
    }

    public GameEntryPoints[] toArray() {
        return (GameEntryPoints[]) super.toArray();
    }

    public ArrayList<String> getNames() {
        final ArrayList<String> temp = new ArrayList<>();
        this.forEach(game -> temp.add(game.getName()));
        return temp;
    }
}

File: Common/Engine/src/common/engine/components/game/GameEntryPoints.kt
package common.engine.components.game

import java.util.*

class GameEntryPoints(vararg gameEntryPoints: GameEntryPoint) : ArrayList<GameEntryPoint>() {

    var activeGame: String? = null
    val names: List<String>
        get() = this.map { game: GameEntryPoint -> game.name }

    init {
        Collections.addAll(this, *gameEntryPoints)
    }

    operator fun get(name: String): GameEntryPoint =
            find { it.name == name } ?: throw IllegalArgumentException("GameEntryPoint not found")

    override operator fun contains(element: GameEntryPoint): Boolean = element.name in this
    operator fun contains(gameName: String): Boolean = this.find { it.name == gameName } != null

    fun render(fps: Int) =
            if (activeGame != null) {
                this[activeGame!!].render(fps)
            } else {
                this.parallelStream().forEach { game ->
                    game.render(fps)
                }
            }

}
